lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -2083,47 +2083,90 @@ void GrandCentralPass::runOnOperation() {
-                    modules.push_back(subMod);
-                    continue;
-                  }
-                  auto subExtMod = cast<FExtModuleOp>(sub);
-                  for (auto anno : AnnotationSet(subExtMod)) {
+
+                DenseSet<FModuleOp> childSet;
+                for (auto path : instancePaths->getAbsolutePaths(mod))
+                  for (InstanceOp instance : path)
+                    childSet.insert(instance->getParentOfType<FModuleOp>());
+
+                // Stop recursiving along this hierarchy if this module is
+                // instantiated by something outside the companion.
+                SmallVector<InstanceRecord *> instances;
+                for (auto *i : instancePaths->instanceGraph.lookup(mod)->uses())
+                  instances.push_back(i);
+
+                // Always modify the output file location of the companion.
+                // Modify the output file location of any modules or external
+                // modules that are instantiated exclusively under the companio.
+                auto isCompanion =
+                    instancePaths->instanceGraph.lookup(mod) == companionNode;
+                if (!isCompanion && !allUnder(instances, companionNode))
+                  continue;
+
+                LLVM_DEBUG({
+                  llvm::dbgs()
+                      << "    - module: " << cast<FModuleLike>(mod).moduleName()
+                      << "\n";
+                });
+
+                FExtModuleOp extmodule = dyn_cast<FExtModuleOp>(mod);
+                if (extmodule) {
+                  for (auto anno : AnnotationSet(extmodule)) {
                     if (!anno.isClass(blackBoxInlineAnnoClass) &&
                         !anno.isClass(blackBoxPathAnnoClass))
                       continue;
-                    if (subExtMod->hasAttr("output_file"))
+                    if (extmodule->hasAttr("output_file"))
                       break;
-                    subExtMod->setAttr(
+                    extmodule->setAttr(
                         "output_file",
                         hw::OutputFileAttr::getAsDirectory(
                             &getContext(),
                             maybeExtractInfo->directory.getValue(),
                             /*excludeFromFileList=*/false));
                     break;
                   }
+                  continue;
+                }
+
+                // Move this module under the Grand Central output directory if
+                // no pre-existing output file information is present.
+                if (!mod->hasAttr("output_file")) {
+                  mod->setAttr("output_file",
+                               hw::OutputFileAttr::getFromDirectoryAndFilename(
+                                   &getContext(),
+                                   maybeExtractInfo->directory.getValue(),
+                                   cast<FModuleLike>(mod).moduleName() + ".sv",
```

Comment:
J/W Do we need to add the file name here? i.e. If we just set the output directory and let ExportVerilog pick the file name?  I realize this would be a bunch of test churn, so no need to do it.

Reply:
I thought this would not work, but it looks like it does! Great catch! I was worried about setting the module name here as this can change underneath us. 😬

Reply:
I'll factor out and land this separately. The change is actually very small and doesn't hit a bunch of tests. Notably, it doesn't affect the end-to-end test (indicating that setting this here is totally unnecessary) only the GCT Views pass test.

Good thinking.